subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/GradleUserHomeTransformerExecutionHistoryRepository.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.transform;
+
+import org.gradle.internal.execution.history.ExecutionHistoryStore;
+
+public class GradleUserHomeTransformerExecutionHistoryRepository extends DefaultTransformerExecutionHistoryRepository {
```

Comment:
I don't really feel comfortable with using the type system to mark things to be different that are not actually different. Not sure what better solution there is to this problem, but wanted to point out nevertheless.

Reply:
We could make the `TransformerInvoker` `Gradle`-scoped. Then it cannot see the mutable transformation workspace provider, since that one is `Project` scoped. That way we would not need the two types which are actually the same any more.

For the future, we need to think about which normalization (especially classpath normalization) should happen for the file inputs of the transformation. It seems like the consumer project would decide that, so we could inject that configuration a `Project` scoped `TransformerInvoker` and we would need to re-introduce the two types we use right now.

What are your thoughts on this?

Reply:
It feels like pinning some of the information related to the project onto `TransformInvoker` is just not a good way to go. For example runtime classpath filtering is now a project-bound thing, but doesn't necessarily have to be so. In fact we chose that approach because it was simplest to implement, but actually, there's nothing project-related about which files you want to ignore when you see a classpath; the setting could be just as well bound to individual tasks. Or source sets. Or whatever. Maybe at some point we'll have settings for how to normalize a runtime classpath come in from all kinds of scopes.

The point is: we shouldn't bind the `Project` to the `TransformInvoker`, but instead let the necessary information be passed to it as needed, from whatever scope it comes from.

Reply:
That sounds like `TransformerInvoker` should be `Gradle` scoped and then we can remove the two types, right?

Reply:
I’m not completely sure which scope we should move it to (what about transforms running in a `SettingsScript` context?), but yeah, I think there should be only one.